pkg/mcs/tso/server/server.go

```diff
@@ -389,83 +394,128 @@ func (s *Server) startGRPCServer(l net.Listener) {
-	case <-time.After(mcsutils.DefaultGRPCGracefulStopTimeout):
-		log.Info("stopping grpc gracefully is taking longer than expected and force stopping now")
-		gs.Stop()
-	}
-
+	log.Info("[tso] grpc server starts serving", zap.String("address", l.Addr().String()))
+	err := s.grpcServer.Serve(l)
 	if s.IsClosed() {
-		log.Info("grpc server stopped")
+		log.Info("[tso] grpc server stopped")
 	} else {
-		log.Fatal("grpc server stopped unexpectedly", errs.ZapError(serverr))
+		log.Fatal("[tso] grpc server stopped unexpectedly", errs.ZapError(err))
 	}
 }
 
 func (s *Server) startHTTPServer(l net.Listener) {
 	defer logutil.LogPanic()
 	defer s.serverLoopWg.Done()
 
-	handler, _ := SetUpRestHandler(s.service)
-	hs := &http.Server{
-		Handler:           handler,
-		ReadTimeout:       5 * time.Minute,
-		ReadHeaderTimeout: 5 * time.Second,
-	}
-	serverr := hs.Serve(l)
-	log.Info("http server stopped serving")
-
-	ctx, cancel := context.WithTimeout(context.Background(), mcsutils.DefaultHTTPGracefulShutdownTimeout)
-	defer cancel()
-	if err := hs.Shutdown(ctx); err != nil {
-		log.Error("http server shutdown encountered problem", errs.ZapError(err))
-	} else {
-		log.Info("all http(s) requests finished")
-	}
+	log.Info("http server starts serving", zap.String("address", l.Addr().String()))
+	err := s.httpServer.Serve(l)
 	if s.IsClosed() {
 		log.Info("http server stopped")
 	} else {
-		log.Fatal("http server stopped unexpectedly", errs.ZapError(serverr))
+		log.Fatal("http server stopped unexpectedly", errs.ZapError(err))
 	}
 }
 
-func (s *Server) startGRPCAndHTTPServers(l net.Listener) {
+func (s *Server) startGRPCAndHTTPServers(serverReadyChan chan<- struct{}, l net.Listener) {
 	defer logutil.LogPanic()
 	defer s.serverLoopWg.Done()
 
 	mux := cmux.New(l)
+	// Don't hang on matcher after closing listener
+	mux.SetReadTimeout(3 * time.Second)
 	grpcL := mux.MatchWithWriters(cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"))
-	httpL := mux.Match(cmux.Any())
+	if s.secure {
+		s.httpListener = mux.Match(cmux.Any())
+	} else {
+		s.httpListener = mux.Match(cmux.HTTP1())
+	}
 
-	s.serverLoopWg.Add(2)
+	s.grpcServer = grpc.NewServer()
+	s.service.RegisterGRPCService(s.grpcServer)
+	diagnosticspb.RegisterDiagnosticsServer(s.grpcServer, s)
+	s.serverLoopWg.Add(1)
 	go s.startGRPCServer(grpcL)
-	go s.startHTTPServer(httpL)
 
+	handler, _ := SetUpRestHandler(s.service)
+	s.httpServer = &http.Server{
+		Handler:     handler,
+		ReadTimeout: 3 * time.Second,
+	}
+	s.serverLoopWg.Add(1)
+	go s.startHTTPServer(s.httpListener)
+
+	serverReadyChan <- struct{}{}
 	if err := mux.Serve(); err != nil {
 		if s.IsClosed() {
-			log.Info("mux stop serving", errs.ZapError(err))
+			log.Info("[tso] mux stopped serving", errs.ZapError(err))
 		} else {
-			log.Panic("mux stop serving unexpectedly", errs.ZapError(err))
+			log.Fatal("[tso] mux stopped serving unexpectedly", errs.ZapError(err))
 		}
 	}
 }
 
+func (s *Server) stopHTTPServer() {
```

Comment:
it is similar to stopGRPCServer. And we could use them in other micro services.

Reply:
Yes, stopHTTPServer and stopGRPCServer are similar, but there is no need to write a general function, because there are still a few differences between these two functions. stopHTTPServer and stopGRPCServer can be used in outher micro services.